File: python/relaymessenger-livekit/src/relaymessenger_livekit/video.py
"""Video in LiveKit's Python shapes, carried by aiortc.

Twin of `packages/livekit/src/video.ts`. Frames are LiveKit's own
`livekit.rtc.VideoFrame` both ways, so a frame read here can go straight to an
`AgentSession` (Gemini Live samples it) and a frame an application already has
for LiveKit can be sent unchanged. The names copy `livekit.rtc`:
`VideoSource.capture_frame(frame, timestamp_us=, rotation=)`,
`LocalVideoTrack.create_video_track(name, source)`, `VideoStream(track)`
yielding `VideoFrameEvent`. aiortc does the encoding (libx264 via PyAV, H.264
constrained baseline ``42e01f``, the profile Cloudflare's SFU accepts) and the
decoding (H.264 and VP8).
"""

from __future__ import annotations

import asyncio
import fractions
import time
from collections import deque
from dataclasses import dataclass
from typing import Any, AsyncIterator, Optional, cast

import av
import numpy as np
from aiortc.mediastreams import MediaStreamError, MediaStreamTrack
from livekit import rtc

VIDEO_TIME_BASE = fractions.Fraction(1, 90_000)

_PACKED_FORMATS: dict[int, tuple[str, int]] = {
    rtc.VideoBufferType.RGBA: ("rgba", 4),
    rtc.VideoBufferType.BGRA: ("bgra", 4),
    rtc.VideoBufferType.ARGB: ("argb", 4),
    rtc.VideoBufferType.ABGR: ("abgr", 4),
    rtc.VideoBufferType.RGB24: ("rgb24", 3),
}


def to_av_frame(frame: rtc.VideoFrame) -> av.VideoFrame:
    """A LiveKit frame as a PyAV frame for aiortc's encoder, without copying through the LiveKit FFI when avoidable."""
    width, height, kind = frame.width, frame.height, frame.type
    data = np.frombuffer(frame.data, dtype=np.uint8)
    if kind == rtc.VideoBufferType.I420 and width % 2 == 0 and height % 2 == 0:
        return av.VideoFrame.from_ndarray(data[: width * height * 3 // 2].reshape(height * 3 // 2, width), format="yuv420p")
    packed = _PACKED_FORMATS.get(kind)
    if packed is not None:
        name, depth = packed
        return av.VideoFrame.from_ndarray(data[: width * height * depth].reshape(height, width, depth), format=name)
    # Anything else (NV12, odd-sized I420, ...) goes through LiveKit's own converter.
    return to_av_frame(frame.convert(rtc.VideoBufferType.RGBA))


def to_rtc_frame(frame: av.VideoFrame) -> rtc.VideoFrame:
    """A decoded PyAV frame as a tightly packed I420 LiveKit frame."""
    if frame.format.name != "yuv420p":
        frame = frame.reformat(format="yuv420p")
    parts = []
    for plane in frame.planes:
        rows = np.frombuffer(plane, dtype=np.uint8).reshape(-1, plane.line_size)
        parts.append(rows[: plane.height, : plane.width].reshape(-1))
    return rtc.VideoFrame(frame.width, frame.height, rtc.VideoBufferType.I420, np.concatenate(parts).tobytes())


@dataclass
class VideoEncoding:
    """LiveKit's ``VideoEncoding``. aiortc adapts its bitrate from REMB; these are recorded for parity."""

    max_bitrate: int = 0
    max_framerate: float = 0


@dataclass
class TrackPublishOptions:
    video_encoding: Optional[VideoEncoding] = None


@dataclass
class RelayVideoSenderStats:
    #: Frames handed to `VideoSource.capture_frame` while published.
    frames_captured: int
    #: Frames aiortc's sender pulled for encoding.
    frames_sent: int
    #: Negotiated codec, for example ``"video/H264"``; ``None`` before the answer.
    codec: Optional[str]
    keyframes: Optional[int] = None
    keyframe_requests: Optional[int] = None


@dataclass
class RelayVideoReceiverStats:
    #: Frames aiortc decoded.
    frames_decoded: int
    #: Frames not delivered because a `VideoStream` was full.
    frames_dropped: int
    codec: Optional[str]
    decode_errors: Optional[int] = None
    keyframe_requests: Optional[int] = None


class VideoSource:
    """The camera feed an application sends (LiveKit ``rtc.VideoSource``)."""

    def __init__(self, width: int, height: int, *, is_screencast: bool = False) -> None:
        self.width = width
        self.height = height
        self.is_screencast = is_screencast
        self._latest: Optional[tuple[rtc.VideoFrame, int]] = None
        self._serial = 0
        self._changed = asyncio.Event()
        self._frames_captured = 0
        self._closed = False

    def capture_frame(
        self,
        frame: rtc.VideoFrame,
        *,
        timestamp_us: int = 0,
        rotation: int = rtc.VideoRotation.VIDEO_ROTATION_0,
    ) -> None:
        """Queue ``frame`` for the next encode; a newer frame replaces one not yet taken, as a live camera does."""
        if self._closed:
            return
        stamp = timestamp_us or int(time.monotonic() * 1_000_000)
        self._latest = (frame, stamp)
        self._serial += 1
        self._frames_captured += 1
        self._changed.set()

    async def aclose(self) -> None:
        self._closed = True
        self._changed.set()

    async def _next(self, after: int) -> tuple[int, rtc.VideoFrame, int]:
        while True:
            if self._closed:
                raise MediaStreamError
            if self._serial > after and self._latest is not None:
                frame, stamp = self._latest
                return self._serial, frame, stamp
            self._changed.clear()
            await self._changed.wait()


class LocalVideoTrack:
    """LiveKit ``rtc.LocalVideoTrack``: a named track over one `VideoSource`."""

    def __init__(self, name: str, source: VideoSource) -> None:
        self.name = name
        self.source = source
        self.kind = "video"

    @staticmethod
    def create_video_track(name: str, source: VideoSource) -> "LocalVideoTrack":
        return LocalVideoTrack(name, source)


class _SenderTrack(MediaStreamTrack):
    """One peer's view of the published `VideoSource`; aiortc's sender pulls it and encodes."""

    kind = "video"

    def __init__(self, sender: "_VideoSender") -> None:
        super().__init__()
        self._sender = sender
        self._serial = 0

    async def recv(self) -> av.VideoFrame:
        if self.readyState != "live":
            raise MediaStreamError
        serial, frame, stamp = await self._sender.source._next(self._serial)
        self._serial = serial
        out = to_av_frame(frame)
        out.pts = stamp * 90_000 // 1_000_000
        out.time_base = VIDEO_TIME_BASE
        self._sender.frames_sent += 1
        return out


class _VideoSender:
    """The published camera across every peer of the call."""

    def __init__(self, track: LocalVideoTrack, options: TrackPublishOptions) -> None:
        self.track = track
        self.source = track.source
        self.options = options
        self.frames_sent = 0
        self.codec: Optional[str] = None
        self.enabled = True
        self._captured_at_publish = track.source._frames_captured

    def create_track(self) -> _SenderTrack:
        return _SenderTrack(self)

    def stats(self) -> RelayVideoSenderStats:
        return RelayVideoSenderStats(
            frames_captured=self.source._frames_captured - self._captured_at_publish,
            frames_sent=self.frames_sent,
            codec=self.codec,
        )


class VideoStream:
    """Frames of a `RemoteVideoTrack` (LiveKit ``rtc.VideoStream``): ``async for event in VideoStream(track)``.

    ``capacity`` 0 keeps every frame, as LiveKit's default does; a positive
    capacity drops the oldest frame when the reader falls behind.
    """

    def __init__(self, track: "RemoteVideoTrack", *, capacity: int = 0) -> None:
        self._track = track
        self._capacity = capacity
        self._queue: deque[Optional[rtc.VideoFrameEvent]] = deque()
        self._ready = asyncio.Event()
        self._closed = False
        track._streams.add(self)
        if track._ended:
            self._end()

    @classmethod
    def from_track(cls, *, track: "RemoteVideoTrack", capacity: int = 0) -> "VideoStream":
        return cls(track, capacity=capacity)

    def _push(self, event: rtc.VideoFrameEvent) -> None:
        if self._closed:
            return
        if self._capacity > 0 and len(self._queue) >= self._capacity:
            self._queue.popleft()
            self._track._frames_dropped += 1
        self._queue.append(event)
        self._ready.set()

    def _end(self) -> None:
        if self._closed:
            return
        self._closed = True
        self._queue.append(None)
        self._ready.set()

    def __aiter__(self) -> AsyncIterator[rtc.VideoFrameEvent]:
        return self

    async def __anext__(self) -> rtc.VideoFrameEvent:
        while not self._queue:
            self._ready.clear()
            await self._ready.wait()
        event = self._queue.popleft()
        if event is None:
            self._queue.append(None)
            raise StopAsyncIteration
        return event

    async def aclose(self) -> None:
        self._track._streams.discard(self)
        self._end()


class RemoteVideoTrack:
    """The other participant's camera (LiveKit ``rtc.RemoteVideoTrack``), one per call, kept across restarts."""

    kind = "video"

    def __init__(self) -> None:
        self._streams: set[VideoStream] = set()
        self._receiver: Optional[asyncio.Task[None]] = None
        self._frames_decoded = 0
        self._frames_dropped = 0
        self._ended = False
        self.codec: Optional[str] = None

    def stats(self) -> RelayVideoReceiverStats:
        return RelayVideoReceiverStats(
            frames_decoded=self._frames_decoded, frames_dropped=self._frames_dropped, codec=self.codec
        )

    def _attach(self, track: MediaStreamTrack) -> None:
        self._detach()
        self._receiver = asyncio.get_running_loop().create_task(self._read(track))

    def _detach(self) -> None:
        if self._receiver is not None:
            self._receiver.cancel()
            self._receiver = None

    def _end(self) -> None:
        self._detach()
        self._ended = True
        for stream in list(self._streams):
            stream._end()

    async def _read(self, track: MediaStreamTrack) -> None:
        # aiortc queues every decoded frame until someone calls recv(), so the
        # track is always drained; frames are converted only for open streams.
        while True:
            try:
                frame: Any = await track.recv()
            except (MediaStreamError, asyncio.CancelledError):
                return
            self._frames_decoded += 1
            if not self._streams or not isinstance(frame, av.VideoFrame):
                continue
            timestamp_us = int(frame.pts * frame.time_base * 1_000_000) if frame.pts is not None and frame.time_base else 0
            event = rtc.VideoFrameEvent(
                frame=to_rtc_frame(frame),
                timestamp_us=timestamp_us,
                rotation=cast(Any, rtc.VideoRotation.VIDEO_ROTATION_0),
            )
            for stream in list(self._streams):
                stream._push(event)

File: packages/sdk/src/calls/video.ts
/**
 * Video tracks for Relay calls, with LiveKit's public API shape.
 *
 * Names and semantics copied from `@livekit/rtc-node` 0.13.35
 * (node_modules/@livekit/rtc-node/src):
 * - video_source.ts:18-75: `new VideoSource(width, height)`,
 *   `captureFrame(frame, timestampUs = BigInt(0), rotation = VIDEO_ROTATION_0)`
 *   which throws "VideoSource is closed" after `close()`.
 * - track.ts:241-268: `LocalVideoTrack.createVideoTrack(name, source)` and
 *   `close(closeSource = true)`.
 * - video_stream.ts:12-16 and :94-98: `VideoFrameEvent { frame, timestampUs,
 *   rotation }` and `class VideoStream extends ReadableStream<VideoFrameEvent>`
 *   built from a remote track, read with `for await`.
 * - participant.ts:737 `publishTrack(track, options)` and :783
 *   `unpublishTrack(...)`; the publish options `videoCodec` and
 *   `videoEncoding { maxBitrate, maxFramerate }` from
 *   @livekit/rtc-ffi-bindings proto/room_pb.d.ts:1792-1859, and `VideoCodec`
 *   (VP8 = 0, H264 = 1) from proto/video_frame_pb.d.ts:28-50.
 * The optional `{ capacity, format }` of `VideoStream` is LiveKit's Python SDK
 * (livekit-rtc/livekit/rtc/video_stream.py:42-60, `RingQueue` in _utils.py:78-94:
 * capacity 0 keeps every frame, a positive capacity drops the oldest).
 *
 * Relay semantics (PROTOCOL.md sections 1-2, _artifacts/calls-facetime-20260922):
 * a participant publishes at most one `video` track; publishing it the first
 * time negotiates it with an add-track offer and announces `userUpdate
 * { video: true }`; unpublishing stops sending and announces `video: false`
 * but leaves the track negotiated, so publishing again only resumes.
 */
import { ReadableStream } from "node:stream/web";
import type { RelayMediaStreamTrackLike } from "./transport.js";
import { VideoBufferType, VideoFrame, VideoRotation } from "./video-frame.js";

export { VideoBufferType, VideoFrame, VideoRotation } from "./video-frame.js";

/** Same names and numbers as LiveKit's `VideoCodec`; Relay sends these two. */
export enum VideoCodec {
  VP8 = 0,
  H264 = 1,
}

/**
 * LiveKit's `VideoEncoding`. A field left out takes LiveKit's camera preset
 * for the frame size being sent (video-presets.ts), for example 3 Mbps at
 * 30 fps for 1920x1080, 1.7 Mbps at 30 fps for 1280x720 and 450 kbps at
 * 20 fps for 640x360.
 */
export interface VideoEncoding {
  /** Encoder target, bits per second. */
  maxBitrate?: number | bigint;
  /** Encoder rate-control hint, frames per second. */
  maxFramerate?: number;
}

/** The subset of LiveKit's `TrackPublishOptions` a Relay call honours. */
export interface TrackPublishOptions {
  /**
   * Preferred codec. Defaults to H.264 (constrained baseline, offered as
   * `profile-level-id=42e01f` with `level-asymmetry-allowed=1` as libwebrtc
   * and LiveKit offer it, sent at the level each frame size needs, 4.0 at
   * 1920x1080 and 30 fps); VP8 is offered second either way.
   */
  videoCodec?: VideoCodec;
  videoEncoding?: VideoEncoding;
}

export type VideoFrameEvent = {
  frame: VideoFrame;
  timestampUs: bigint;
  rotation: VideoRotation;
};

/** Counters for the published video track; timestamps are epoch ms. */
export interface RelayVideoSenderStats {
  /** Negotiated send codec, learned from the answer; undefined before it. */
  codec: "h264" | "vp8" | undefined;
  framesCaptured: number;
  framesEncoded: number;
  /** Captured while unpublished, before the codec was known, or while the encoder was behind. */
  framesDropped: number;
  keyframes: number;
  rtpPackets: number;
  bytes: number;
  /** PLI and FIR received from the SFU. */
  keyframeRequests: number;
  encodeErrors: number;
  firstFrameAt: number | undefined;
  lastFrameAt: number | undefined;
  /** Frames encoded in the last 5 s. */
  recentFrames: number;
  width: number | undefined;
  height: number | undefined;
}

/** Counters for the other participant's video track; timestamps are epoch ms. */
export interface RelayVideoReceiverStats {
  codec: string | undefined;
  rtpPackets: number;
  /** Whole frames assembled from RTP. */
  framesAssembled: number;
  /** Frames dropped because a packet was lost or malformed, or a delta arrived before a keyframe. */
  framesDropped: number;
  framesDecoded: number;
  decodeErrors: number;
  /** PLI sent to ask the sender for a keyframe. */
  keyframeRequests: number;
  firstFrameAt: number | undefined;
  lastFrameAt: number | undefined;
  /** Frames decoded in the last 5 s. */
  recentFrames: number;
  width: number | undefined;
  height: number | undefined;
}

/** @internal An engine's encoder and RTP writer for the one local video track. */
export interface RelayVideoSenderLike {
  /** The engine track, reused by every peer connection of the call. */
  readonly track: RelayMediaStreamTrackLike;
  /** A new peer added `track`; bind before its offer is created. */
  bind(transceiver: unknown, peer: unknown): void;
  capture(frame: VideoFrame, timestampUs: bigint, rotation: VideoRotation): void;
  setEnabled(enabled: boolean): void;
  stats(): RelayVideoSenderStats;
  close(): void;
}

/** @internal An engine's RTP reader and decoder for one pulled video track. */
export interface RelayVideoReceiverLike {
  onframe: ((event: VideoFrameEvent) => void) | null;
  /** Decode only while a `VideoStream` reads; `true` also asks the sender for a keyframe. */
  setActive(active: boolean): void;
  stats(): RelayVideoReceiverStats;
  stop(): void;
}

/** @internal */
export interface RelayVideoFactory {
  createVideoSender(options: TrackPublishOptions): Promise<RelayVideoSenderLike>;
  createVideoReceiver(track: RelayMediaStreamTrackLike, transceiver: unknown): RelayVideoReceiverLike;
}

type FrameSink = (frame: VideoFrame, timestampUs: bigint, rotation: VideoRotation) => void;

export class VideoSource {
  width: number;
  height: number;
  /** @internal */
  closed = false;
  readonly #sinks = new Set<FrameSink>();

  constructor(width: number, height: number) {
    this.width = width;
    this.height = height;
  }

  /**
   * Hand one raw frame to the published track. RGBA, BGRA and I420 (LiveKit
   * `VideoBufferType`) go to the encoder as they are; the other packed RGB
   * layouts are converted to I420 first. `timestampUs` 0 means "now".
   * Frames captured while the track is not published are dropped.
   */
  captureFrame(frame: VideoFrame, timestampUs = BigInt(0), rotation = VideoRotation.VIDEO_ROTATION_0): void {
    if (this.closed) throw new Error("VideoSource is closed");
    for (const sink of this.#sinks) sink(frame, timestampUs, rotation);
  }

  async close(): Promise<void> {
    this.closed = true;
    this.#sinks.clear();
  }

  /** @internal */
  _attach(sink: FrameSink): () => void {
    this.#sinks.add(sink);
    return () => this.#sinks.delete(sink);
  }
}

export class LocalVideoTrack {
  readonly kind = "video" as const;
  readonly name: string;
  /** @internal */
  readonly source: VideoSource;

  constructor(name: string, source: VideoSource) {
    this.name = name;
    this.source = source;
  }

  static createVideoTrack(name: string, source: VideoSource): LocalVideoTrack {
    return new LocalVideoTrack(name, source);
  }

  async close(closeSource = true): Promise<void> {
    if (closeSource) await this.source.close();
  }
}

interface StreamConsumer {
  push(event: VideoFrameEvent): void;
  end(): void;
}

/**
 * The other participant's camera. Emitted once per call by the transport's
 * `trackSubscribed` event; it survives SFU restarts (the transport moves it
 * onto the new session's receiver), so a `VideoStream` read from it keeps going.
 */
export class RemoteVideoTrack {
  readonly kind = "video" as const;
  readonly name = "video";
  #receiver: RelayVideoReceiverLike | undefined;
  #retired: RelayVideoReceiverStats | undefined;
  readonly #consumers = new Set<StreamConsumer>();
  #ended = false;

  /** Counters summed over every receiver this track has had. */
  stats(): RelayVideoReceiverStats | undefined {
    const current = this.#receiver?.stats();
    const retired = this.#retired;
    if (!retired) return current;
    if (!current) return { ...retired, recentFrames: 0 };
    return {
      ...current,
      rtpPackets: retired.rtpPackets + current.rtpPackets,
      framesAssembled: retired.framesAssembled + current.framesAssembled,
      framesDropped: retired.framesDropped + current.framesDropped,
      framesDecoded: retired.framesDecoded + current.framesDecoded,
      decodeErrors: retired.decodeErrors + current.decodeErrors,
      keyframeRequests: retired.keyframeRequests + current.keyframeRequests,
      firstFrameAt: retired.firstFrameAt ?? current.firstFrameAt,
      lastFrameAt: current.lastFrameAt ?? retired.lastFrameAt,
    };
  }

  /** @internal The transport moves the track onto a new session's receiver. */
  _attach(receiver: RelayVideoReceiverLike): void {
    this._detach();
    this.#receiver = receiver;
    receiver.onframe = (event) => {
      for (const consumer of this.#consumers) consumer.push(event);
    };
    receiver.setActive(this.#consumers.size > 0);
  }

  /** @internal */
  _detach(): void {
    const receiver = this.#receiver;
    if (!receiver) return;
    this.#receiver = undefined;
    const stats = receiver.stats();
    const retired = this.#retired;
    this.#retired = retired
      ? {
        ...stats,
        rtpPackets: retired.rtpPackets + stats.rtpPackets,
        framesAssembled: retired.framesAssembled + stats.framesAssembled,
        framesDropped: retired.framesDropped + stats.framesDropped,
        framesDecoded: retired.framesDecoded + stats.framesDecoded,
        decodeErrors: retired.decodeErrors + stats.decodeErrors,
        keyframeRequests: retired.keyframeRequests + stats.keyframeRequests,
        firstFrameAt: retired.firstFrameAt ?? stats.firstFrameAt,
        lastFrameAt: stats.lastFrameAt ?? retired.lastFrameAt,
      }
      : stats;
    receiver.onframe = null;
    receiver.stop();
  }

  /** @internal The call is over: every `VideoStream` ends. */
  _end(): void {
    this._detach();
    this.#ended = true;
    for (const consumer of [...this.#consumers]) consumer.end();
    this.#consumers.clear();
  }

  /** @internal */
  _subscribe(consumer: StreamConsumer): () => void {
    if (this.#ended) {
      consumer.end();
      return () => undefined;
    }
    this.#consumers.add(consumer);
    if (this.#consumers.size === 1) this.#receiver?.setActive(true);
    return () => {
      this.#consumers.delete(consumer);
      if (this.#consumers.size === 0) this.#receiver?.setActive(false);
    };
  }
}

export interface VideoStreamOptions {
  /** Frames kept for a slow reader; 0 (the default) keeps all, a positive number drops the oldest. */
  capacity?: number;
  /** Pixel layout of every frame read; defaults to the decoder's I420. */
  format?: VideoBufferType;
}

/**
 * Decoded frames of a remote video track. Decoding starts when the first
 * stream opens and stops when the last one is cancelled.
 */
export class VideoStream extends ReadableStream<VideoFrameEvent> {
  constructor(track: RemoteVideoTrack, options: VideoStreamOptions = {}) {
    const capacity = options.capacity ?? 0;
    const format = options.format;
    const queue: VideoFrameEvent[] = [];
    let waiting: ((value: void) => void) | undefined;
    let ended = false;
    let unsubscribe: (() => void) | undefined;
    const wake = (): void => {
      const resolve = waiting;
      waiting = undefined;
      resolve?.();
    };
    super({
      start: () => {
        unsubscribe = track._subscribe({
          push: (event) => {
            if (capacity > 0 && queue.length >= capacity) queue.shift();
            queue.push(event);
            wake();
          },
          end: () => {
            ended = true;
            wake();
          },
        });
      },
      pull: async (controller) => {
        while (queue.length === 0 && !ended) await new Promise<void>((resolve) => { waiting = resolve; });
        const event = queue.shift();
        if (!event) {
          controller.close();
          return;
        }
        const frame = format === undefined || format === event.frame.type ? event.frame : event.frame.convert(format);
        controller.enqueue({ ...event, frame });
      },
      cancel: () => {
        unsubscribe?.();
        ended = true;
        wake();
      },
    }, { highWaterMark: 0 });
  }
}
